i3lock: add module

offlinehacker opened the pull request.

offlinehacker commented:
This PR dependens on: https://github.com/rycee/home-manager/pull/718

uvNikita commented:
Looks like a very comprehensive module! I'm a user of i3lock-color myself, so I'll look at it and test as soon as I'll have time. 

Thank you for the PR!

uvNikita commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,291 @@
+  };
+
+  config = mkIf cfg.enable {
+    xsession.lockCmd =
```
I'm probably missing something, but why we can't just enable screen-locker service and set `lockCmd` there? It was designed to be used with any locks, it just integrates them with systemd (loginctl).

```nix
services.screen-locker = {
  enbale = true;
  lockCmd = ...
};
```

teto replied:
loginctl already abstracts the locking mechanism from what I read so seems like a good approach

offlinehacker replied:
Agreed, thx, changed.

uvNikita commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,291 @@
+      "--datefont=${cfg.clock.dateFont}")
+
+    # text
+    (optionalString (cfg.textSize != null) "--datefont=${cfg.textSize}")
```
You probably wanted to use this options from i3lock-color(1):

> --{time, date, layout, verif, wrong}size=number
              Sets the font size used to render various strings.

offlinehacker replied:
Implemented said options

uvNikita commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,291 @@
+      };
+    };
+
+    verifyText = mkOption {
```
All the "text" options can probably be grouped under the `i3lock.text` submodule:

```nix
i3lock.text = {
  verify = "...";
  wrong = "...";
  size = 16;
};
```

offlinehacker replied:
Implemented

uvNikita commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,291 @@
+      default = null;
+    };
+
+    modSize = mkOption {
```
Doesn't seem like this option is used anywhere and I can't find it in manual.

offlinehacker replied:
Removed

uvNikita commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,291 @@
+      description = "Colors to use for i3lock (uses i3lock-color fork instead of i3lock).";
+      type = types.nullOr (types.submodule {
+        options = {
+          insidevercolor = mkOption {
```
Can probably drop the "color" suffix here since these options are already inside the `color` submodule.

offlinehacker replied:
Fixed

offlinehacker commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,291 @@
+      default = if enablei3LockColor then pkgs.i3lock-color else pkgs.i3lock;
+    };
+
+    color = mkOption {
```
Implemented

offlinehacker commented:
@uvNikita thanks for in depth review, i implemented your suggestions and fixes and also added a few other extra options, which should cover most of i3lock-color options. Additionally i changed default package to `i3lock-color`, as this is what this module actually implements, but it can be still used with normal `i3lock` in case if you do not use color related options.

offlinehacker commented:
Any updates on getting this merged?

uvNikita commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,412 @@
+        ''${pkgs.scrot}/bin/scrot /tmp/screen_locked.png
+        ''${pkgs.imagemagick}/bin/convert /tmp/screen_locked.png \
+          -scale 10% -scale 1000% /tmp/screen_locked.png
+      '';
```
This example seems to render incorrectly in the manual:

```
programs.i3lock.extraCommands
           Extra commands to run before running i3lock.

           Type: strings concatenated with "\n"

           Default: ""

           Example:

               ''
               ''${pkgs.scrot}/bin/scrot /tmp/screen_locked.png
               ''${pkgs.imagemagick}/bin/convert /tmp/screen_locked.png \
                 -scale 10% -scale 1000% /tmp/screen_locked.png
               ''
```

Maybe we should drop the `''` before $?

offlinehacker replied:
Fixed using `literalExample`.

uvNikita commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,412 @@
+      };
+    };
+
+    extraCommands = mkOption {
```
Do you think it makes sense to have two options: `extraCommandsBefore` and `extraCommandsAfter`? E.g. I've been using the following locking script to make sure there is only US layout when i3lock is working and reseting it to the default configuration after successful unlock:

```
  locker = pkgs.writeShellScriptBin "lock" ''
    ${pkgs.xorg.setxkbmap}/bin/setxkbmap us
    ${pkgs.i3lock-color}/bin/i3lock-color --nofork
    ${pkgs.systemd}/bin/systemctl --user restart setxkbmap.service
  '';
```
Note `--nofork` otion that should be used in this case.

If you think this is too special case to cover in this module, we can leave it as is :)

offlinehacker replied:
Makes sense, added suggested `extraCommandBefore` and `extraCommandsAfter`.

offlinehacker commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,412 @@
+      enable = mkEnableOption "show clock on i3lock.";
+
+      timeStr = mkOption {
+        description = "i3lock clock time format string.";
```
Makes sense

offlinehacker commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,412 @@
+        default = "%H:%M:%S";
+      };
+
+      dateStr = mkOption {
```
Agreed

offlinehacker replied:
Agreed

offlinehacker commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,412 @@
+    clock = {
+      enable = mkEnableOption "show clock on i3lock.";
+
+      timeStr = mkOption {
```
Agreed

offlinehacker commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,412 @@
+  ]);
+
+  lockScript = pkgs.writeScript "i3lock.sh" ''
+    #!${pkgs.stdenv.shell}
```
Didn't knew, thanks!

offlinehacker commented on modules/programs/i3lock.nix:
```diff
@@ -0,0 +1,412 @@
+        default = null;
+      };
+
+      bshl = mkOption {
```
Yup

offlinehacker commented:
Implemented suggested changes, thanks!

JustinLovinger commented:
I found an error in this module. If you set `programs.i3lock.clock = enable;`, the `xss-lock.service` will give an error like "Failed to resolve unit specifiers in --timestr=%H:%M:%S: Invalid slot". `lockCmd` should be turned into a script, even if there are no extra commands, and the script used for `services.screen-locker.lockCmd`, instead of passing the command string itself.

Also, maybe values should be quoted, like `--datestr="${cfg.clock.dateFormat}"`, instead of `--datestr=${cfg.clock.dateFormat}`.

offlinehacker commented:
I rebased and implemented suggested changes. @JustinLovinger can you test it agan, since i switched to sway and cannot test it currently?

offlinehacker commented:
I implemented additional validation checks for colors and escaping for all text args.

JustinLovinger commented:
@offlinehacker  `clock` works now, but another error was introduced. `background.color` should actually be a 6 character `rrggbb` string, but the type checks for an 8 character string. `i3lock` background doesn't have a transparency value. Also, the error message when a check fails is not very helpful. The error looks like

```
error: The option value `home-manager.users.<name>.programs.i3lock.background.color' in `path/to/file.nix' is not of type `null or string'.
```

DamienCassou commented:
Any update on this?
